sql, distsql: fix mismatches between needed and plan columns in opt_needed

heyihong opened the pull request:
[The comment](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/data_source.go#L119-L121) mentions that sourceColumns match the plan.Columns() 1-to-1. But current code doesn't satisfy this assumption. It seems that the length of sourceColumns may not be equal to the length plan.Columns(). If that's the case, for fixing such mismatch, we should use plan.Columns() to calculate sourceNeeded.

Release note (bug fix): fixed mismatches between needed and plan columns
in opt_needed for joinNode, filterNode and renderNode

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/21696)
<!-- Reviewable:end -->

RaduBerinde commented:
@knz is the mismach between `sourceColumns` and `plan.Columns()` in these cases intentional, or is it a bug? If it's intentional, please suggest an updated comment for `sourceColumns`.

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21696#-:-L3_x8doDIpkLKLvjxxg:bwpa1um)*
<!-- Sent from Reviewable.io -->

heyihong commented:
For joinNode, it seems that we need to trim leftNeeded and rightNeeded. Because I found mismatch in https://teamcity.cockroachdb.com/viewLog.html?buildId=490677&tab=buildResultsDiv&buildTypeId=Cockroach_ExampleORMs

heyihong commented:
As far as I know, there are multiple place that may cause such mismatch:
1. [inserting render node in distinct](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/distinct.go#L173-L182)
2. [sortNode elision](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/expand_plan.go#L587)
3. [having filterNode](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/group.go#L268)

heyihong committed:
sql, distsql: fix mismatches between needed and plan columns in opt_needed

The length of needed and planColumns are not equal due to not using
len(planColumns(child plan)) to create source needed array.

Release note (bug fix): fixed mismatches between needed and plan columns
in opt_needed for joinNode, filterNode and renderNode

heyihong commented:
Ping

RaduBerinde commented:
@knz can you take a look at my question above?

knz commented:
So I looked and I think the mismatch was not intentional. I distinctly recall my intent that sourceColumns matches the source plan's columns! 

The reason why this was not achieved, I think, is that we didn't have a clear picture in our tests that the code was not matching intent (i.e. it didn't matter in practice). I'm not sure whether the original intent matters, but it would make understanding this code, and future reviews, easier to align them.

heyihong commented:
Please take a look

RaduBerinde commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/sql/opt_needed.go, line 126 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21696#-L4mK2pnGpUvaG7Gpi1f:-L4mK2pnGpUvaG7Gpi1g:byd63dq) ([raw file](https://github.com/cockroachdb/cockroach/blob/894745b995e865ae8662113d8795959f304cf5bb/pkg/sql/opt_needed.go#L126)):*
> ```Go
> 		sourceNeeded := make([]bool, len(planColumns(n.source.plan)))
> 		copy(sourceNeeded, needed)
> 		for i := 0; i < len(n.source.info.sourceColumns) && i < len(sourceNeeded); i++ {
> ```

This is very confusing. If they have different lengths, what's `i`? Is it the `i`th sourceColumn or is it the `i`th plan column? We are assuming that the shorter one maps 1:1 to the prefix of the longer one.

I think we should fix the cases where these are different instead of tolerating this mismatch.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21696)*
<!-- Sent from Reviewable.io -->

heyihong commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/sql/opt_needed.go, line 126 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21696#-L4mK2pnGpUvaG7Gpi1f:-L4mukjGzhFsAf5mVJX4:b-72j1za) ([raw file](https://github.com/cockroachdb/cockroach/blob/894745b995e865ae8662113d8795959f304cf5bb/pkg/sql/opt_needed.go#L126)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This is very confusing. If they have different lengths, what's `i`? Is it the `i`th sourceColumn or is it the `i`th plan column? We are assuming that the shorter one maps 1:1 to the prefix of the longer one.

I think we should fix the cases where these are different instead of tolerating this mismatch.
</blockquote></details>

I think the assumption may not be the original intent, but it works on the current both. Since I found that there are three possible relations between len(n.source.info.sourceColumns) and len(planColumns(n.source.plan)): <, =, > for renderNode. 

It is possible that len(n.source.info.sourceColumns) > len(planColumns(n.source.plan)). For example, for the sql below len(planColumns(n.source.plan)) = 0 while len(n.source.info.sourceColumns) > 0.
```sql
tatement ok
CREATE SEQUENCE select_test

query IIB colnames
SELECT * FROM select_test
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21696)*
<!-- Sent from Reviewable.io -->

heyihong commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/sql/opt_needed.go, line 126 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21696#-L4mK2pnGpUvaG7Gpi1f:-L4n6M6dgJKb45PKu8li:b-nhvy0k) ([raw file](https://github.com/cockroachdb/cockroach/blob/894745b995e865ae8662113d8795959f304cf5bb/pkg/sql/opt_needed.go#L126)):*
<details><summary><i>Previously, heyihong (Yihong He) wrote…</i></summary><blockquote>

I think the assumption may not be the original intent, but it works on the current both. Since I found that there are three possible relations between len(n.source.info.sourceColumns) and len(planColumns(n.source.plan)): <, =, > for renderNode. 

It is possible that len(n.source.info.sourceColumns) > len(planColumns(n.source.plan)). For example, for the sql below len(planColumns(n.source.plan)) = 0 while len(n.source.info.sourceColumns) > 0.
```sql
tatement ok
CREATE SEQUENCE select_test

query IIB colnames
SELECT * FROM select_test
```
</blockquote></details>

Maybe I should fix the the planColumns of sequenceSelectNode instead of doing this

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21696)*
<!-- Sent from Reviewable.io -->

heyihong commented:
Also, `zeroNode` may also lead to len(n.source.info.sourceColumns) > len(planColumns(n.source.plan))

RaduBerinde commented:
*[pkg/sql/opt_needed.go, line 126 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21696#-L4mK2pnGpUvaG7Gpi1f:-L4pz82JoWAYabuykjTv:banbjv) ([raw file](https://github.com/cockroachdb/cockroach/blob/894745b995e865ae8662113d8795959f304cf5bb/pkg/sql/opt_needed.go#L126)):*
<details><summary><i>Previously, heyihong (Yihong He) wrote…</i></summary><blockquote>

Maybe I should fix the the planColumns of sequenceSelectNode instead of doing this
</blockquote></details>

Right, that's what @knz and I were saying - we should fix the `sourceColumns` that are incorrect instead of working around the mismatch.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21696)*
<!-- Sent from Reviewable.io -->

knz commented:
We have removed the heuristic planner so this entire area of the code has gone away. Therefore, we choose to not pursue this code change further.
